src/tests/GraphTest.hpp

```diff
@@ -64,54 +69,49 @@ class GraphTest : public Test {
             
             /* set interval to inf so this won't be killed by timeout */
             gtc->interval = numeric_limits<double>::max();
+            auto stats = g->grab_stats();
+            std::apply(print_stats, stats);
+            insertionProb = 3300;
+            removalProb = 3300;
+            lookupProb = 1650;
+            clearProb = 1650;
+            workingThreads = gtc->task_num;
+            threadsDone = 0;
         }
 
         int execute(GlobalTestConfig *gtc, LocalTestConfig *ltc) {
             int tid = ltc->tid;
             std::mt19937_64 gen_p(ltc->seed);
             std::mt19937_64 gen_v(ltc->seed + 1);
-            for (size_t i = 0; i < thd_ops[ltc->tid]; i++) {
-                unsigned p = gen_p()%100;
-                int src = gen_v() % max_verts;
-                int dest = gen_v() % max_verts;
-                if (p<prop_inserts) {
-                    //std::cout << "Insert(" << src << ", " << dest << ")" << std::endl;
-                    g->add_edge(src, dest, 1);
-                } else if (p<prop_removes) {
-                    //std::cout << "Delete(" << src << ", " << dest << ")" << std::endl;
-                    g->for_each_edge(src, [&dest](int v) { dest = v; return false; });
-                    g->remove_edge(src, dest);
-                } else if (p < prop_lookup) {
-                    //std::cout << "Lookup(" << src << "," << dest << ")" << std::endl;
-                    g->has_edge(src, dest);
+            std::uniform_int_distribution<> dist(0,9999);
+            std::uniform_int_distribution<> distv(0,max_verts-1);
+            int rng = dist(gen_p);
+            for (size_t i = 0; i < thd_ops[tid]; i++) {
+                if (rng <= insertionProb) {
+                    // std::cout << "rng(" << rng << ") is add_edge <= " << insertionProb << std::endl; 
+                    g->add_edge(distv(gen_v), distv(gen_v), -1);
+                } else if (rng <= insertionProb + removalProb) {
+                    // std::cout << "rng(" << rng << ") is remove_any_edge <= " << insertionProb + removalProb << std::endl; 
+                    g->remove_edge(distv(gen_v), distv(gen_v));
```

Comment:
The probability of successfully adding a new edge (i.e., P(succeeds in adding an edge | adding an edge) is higher than successfully removing an existing edge, I think. Say at the beginning each vertex has 128 edges on average and there're 1048576 vertex slots and 524288 existing vertices, then the probability of successfully adding an edge is 1/2*((524288-128)/1048575), while the probability of successfully removing an edge is 1/2*(128/1048575). The latter is significantly less than the former, which means the average number of edges are increasing during the execution.

Reply:
:thinking:

Reply:
FWIW, I do think this issue also appears in ChurnTest for maps and sets. It's normally not a problem when we set prefill the data structure to half full.

Reply:
Actually I think these are different: successfully adding and removing a node from map/set both have probability of 1/2 if the container is filled by half. But as I said, successfully adding and removing an edge have different probability: the former is 1/2*((524288-128)/1048575) but the latter is 1/2*(128/1048575). The problem is the that _edges for each vertex_ is not filled in half

Reply:
To be clear: filling edges by half means 1/2 of existing vertices neighbors every vertex.

Reply:
Just did a quick calculation: if in total there are 1048575 (1024*1024) vertices, on average half of them exist, and each vertex has 128 edges on average, then add_edge:remove_edge ratio has to be 1/4096 to keep the average number of edges of each vertex around 128.

Reply:
That's indeed too many edges given the vertex number. Let's go with a pre-set smaller ratio then.

Reply:
Does this also mean that the possibility of _actually_ deleting/adding an edge is pretty low? Do you want to consider a smaller graph?

Reply:
I've pushed my solution in commit [59a9723](https://github.com/urcs-sync/Montage/pull/23/commits/59a9723967a627f41ca4d0c8b7ed5fc8813235d6). See what you think.

Reply:
Here are some results from newer changes...

```
|V|=528594,|E|=68210682,average degree = 129.042,maximum degree = 327
|V|=304841,|E|=17752737,average degree = 58.2361,maximum degree = 5339
add_edge = 15 (0.00339982%)
, remove_edge = 96 (0.0217588%)
, add_vertex = 108668 (24.6301%)
, remove_vertex = 332421 (75.3447%)
```

Even with varying numbers of tasks, the ratio is about the same here. The average degree drops pretty sharply, here, which is due to `remove_vertex` succeeding 3x more often than `add_vertex`.

Reply:
Notice that the average degree stays stable when `add_edge:remove_edge=3:1`. I don't understand why but as an ad-hoc solution I just pushed it to the PR. Trying to find out where this magical 3:1 comes from for another 30 minutes and give up afterwards.